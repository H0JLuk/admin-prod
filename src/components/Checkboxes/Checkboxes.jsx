import React, { useCallback } from 'react';
import { Checkbox } from 'antd';

import styles from './Checkboxes.module.css';

const ALL_APPS_CHECKBOX_TEXT = 'Все приложения';

const CheckBoxCustom = ({ label, name, disabled, onChange, checked }) => {
    const onChangeHandler = useCallback((e) => onChange(e.target.checked, name), [onChange, name]);

    return (
        <Checkbox
            className={ styles.checkbox }
            name={ name }
            onChange={ onChangeHandler }
            checked={ checked }
            disabled={ disabled }
        >
            { label }
        </Checkbox>
    );
};

const Checkboxes = ({ checkboxesData = {}, onChange = () => {}, onChangeAll = () => {}, disabledAll = false }) => {
    const checkboxesKeys = Object.keys(checkboxesData);
    const allAreChecked = checkboxesKeys.every((key) => checkboxesData[key].checked);

    return (
        <>
            <div className={ styles.checkboxAllApps }>
                <CheckBoxCustom
                    label={ ALL_APPS_CHECKBOX_TEXT }
                    disabled={ disabledAll }
                    onChange={ onChangeAll }
                    checked={ allAreChecked }
                    name="all"
                />
            </div>
            <div className={ styles.checkbox }>
                { checkboxesKeys.map((key) => {
                    const { label, disabled, checked } = checkboxesData[key];

                    return (
                        <CheckBoxCustom
                            name={ key }
                            label={ label }
                            disabled={ disabledAll || disabled }
                            onChange={ onChange }
                            key={ key }
                            checked={ checked }
                        />
                    );
                }) }
            </div>
        </>
    );
};

export default Checkboxes;
