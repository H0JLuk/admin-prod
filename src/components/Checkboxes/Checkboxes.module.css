.checkbox {
    display: block;
    margin: 0 !important;
    display: flex;
    flex-wrap: wrap;
    align-items: center;
}

.checkbox > label {
    margin-bottom: 18px !important;
}

.checkbox :global(.ant-checkbox-inner) {
    width: 18px;
    height: 18px;
    outline: none;
}

.checkbox :global(.ant-checkbox-checked .ant-checkbox-inner) {
    background-color: #09a552;
    border-color: #d9d9d9;
}

.checkbox :global(.ant-checkbox-input:focus + .ant-checkbox-inner) {
    border-color: #d9d9d9;
}

.checkbox :global(.ant-checkbox-input:hover + .ant-checkbox-inner) {
    border-color: #09a552;
}

.checkbox :global(.ant-checkbox-checked .ant-checkbox-inner)::after {
    width: 5px;
    height: 9px;
}

.checkbox :global(.ant-checkbox-disabled.ant-checkbox-checked .ant-checkbox-inner::after) {
    border-color: #fff;
    left: 25% !important;
}

.checkbox :global(.ant-checkbox:hover::after),
.checkbox :global(.ant-checkbox-wrapper:hover .ant-checkbox::after) {
    visibility: hidden;
}

.checkbox :global(.ant-checkbox-checked + span) {
    color: #494949;
}

.checkbox :global(.ant-checkbox-wrapper) {
    flex-basis: 33%;
    display: flex;
    align-items: center;
    flex-wrap: nowrap;
}

.checkboxAllApps :global(.ant-checkbox-wrapper) {
    flex-wrap: nowrap;
}

.checkboxAllApps {
    flex-basis: 20%;
    min-width: 135px;
    margin-right: 10px;
    flex-shrink: 0;
}

.checkboxAllApps > label > span ~ span {
    min-width: 140px;
}
