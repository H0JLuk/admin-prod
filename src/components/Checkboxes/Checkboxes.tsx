import React, { useCallback } from 'react';
import { Checkbox } from 'antd';
import { CheckboxChangeEvent } from 'antd/lib/checkbox';
import noop from 'lodash/noop';

import styles from './Checkboxes.module.css';

export const ALL_APPS_CHECKBOX_TEXT = 'Все приложения';

type CheckboxesData = {
    id?: number;
    label: string;
    disabled: boolean;
    checked: boolean;
};

type CheckBoxCustomProps = CheckboxesData & {
    name: string;
    onChange: (checked: boolean, name: string) => void;
};

type CheckboxesProps = {
    checkboxesData: Record<string, CheckboxesData>;
    onChange: (checked: boolean, name: string) => void;
    onChangeAll: (checked: boolean) => void;
    disabledAll: boolean;
};

export const CheckBoxCustom: React.FC<CheckBoxCustomProps> = ({
    label,
    name,
    disabled,
    onChange,
    checked,
}) => {
    const onChangeHandler = useCallback((e: CheckboxChangeEvent) => onChange(e.target.checked, name), [onChange, name]);

    return (
        <Checkbox
            className={styles.checkbox}
            name={name}
            onChange={onChangeHandler}
            checked={checked}
            disabled={disabled}
        >
            {label}
        </Checkbox>
    );
};

const Checkboxes: React.FC<CheckboxesProps> = ({
    checkboxesData = {},
    onChange = noop,
    onChangeAll = noop,
    disabledAll = false,
}) => {
    const checkboxesKeys = Object.keys(checkboxesData);
    const allAreChecked = checkboxesKeys.every((key) => checkboxesData[key].checked);

    return (
        <>
            <div className={styles.checkboxAllApps}>
                <CheckBoxCustom
                    label={ALL_APPS_CHECKBOX_TEXT}
                    disabled={disabledAll}
                    onChange={onChangeAll}
                    checked={allAreChecked}
                    name="all"
                />
            </div>
            <div className={styles.checkbox}>
                {checkboxesKeys.map((key) => {
                    const { label, disabled, checked } = checkboxesData[key];

                    return (
                        <CheckBoxCustom
                            name={key}
                            label={label}
                            disabled={disabledAll || disabled}
                            onChange={onChange}
                            key={key}
                            checked={checked}
                        />
                    );
                })}
            </div>
        </>
    );
};

export default Checkboxes;
