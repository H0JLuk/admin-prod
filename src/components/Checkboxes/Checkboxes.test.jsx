import React from 'react';
import { shallow } from 'enzyme';
import Checkboxes, { CheckBoxCustom } from './Checkboxes';

const TEST_PROPS_CHECKBOX_CUSTOM = [
    {
        label: 'TEST_LABEL_ONE',
        disabled: false,
        checked: false,
        name: 'TEST_KEY_ONE',
        onChange: jest.fn(),
    },
    {
        label: 'TEST_LABEL_TWO',
        disabled: false,
        checked: false,
        name: 'TEST_KEY_TWO',
        onChange: jest.fn(),
    },
];

const TEST_PROPS = {
    checkboxesData: {
        TEST_KEY_ONE: TEST_PROPS_CHECKBOX_CUSTOM[0],
        TEST_KEY_TWO: TEST_PROPS_CHECKBOX_CUSTOM[1],
    },
    onChange: jest.fn(),
    onChangeAll: jest.fn(),
    disabledAll: false,
};

describe('<Checkboxes /> test', () => {
    const Component = shallow(<Checkboxes { ...TEST_PROPS } />);

    it('should be mount snap', () => {
        expect(Component.html()).toMatchSnapshot();
    });

    it('should call onChangeAll function', () => {
        const CheckboxCheckAll = Component.find('[name="all"]');
        CheckboxCheckAll.simulate('change');
        expect(TEST_PROPS.onChangeAll).toBeCalledTimes(1);
    });

    it('should checked one checkbox', () => {
        const custom = shallow(<CheckBoxCustom { ...TEST_PROPS_CHECKBOX_CUSTOM[0] } />);
        custom.simulate('change', { target: { checked: true } });
        expect(TEST_PROPS_CHECKBOX_CUSTOM[0].onChange).toHaveBeenCalledWith(true, 'TEST_KEY_ONE');
        expect(TEST_PROPS_CHECKBOX_CUSTOM[0].onChange).toBeCalledTimes(1);
    });
});
