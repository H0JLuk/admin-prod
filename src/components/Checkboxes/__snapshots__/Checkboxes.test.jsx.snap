// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkboxes /> test should be mount snap 1`] = `"<div class=\\"checkboxAllApps\\"><label class=\\"ant-checkbox-wrapper checkbox\\"><span class=\\"ant-checkbox\\"><input type=\\"checkbox\\" name=\\"all\\" class=\\"ant-checkbox-input\\"/><span class=\\"ant-checkbox-inner\\"></span></span><span>Все приложения</span></label></div><div class=\\"checkbox\\"><label class=\\"ant-checkbox-wrapper checkbox\\"><span class=\\"ant-checkbox\\"><input type=\\"checkbox\\" name=\\"TEST_KEY_ONE\\" class=\\"ant-checkbox-input\\"/><span class=\\"ant-checkbox-inner\\"></span></span><span>TEST_LABEL_ONE</span></label><label class=\\"ant-checkbox-wrapper checkbox\\"><span class=\\"ant-checkbox\\"><input type=\\"checkbox\\" name=\\"TEST_KEY_TWO\\" class=\\"ant-checkbox-input\\"/><span class=\\"ant-checkbox-inner\\"></span></span><span>TEST_LABEL_TWO</span></label></div>"`;
